feat: split raw data storage into per-dataset SQLite files (#8219)
microsoft/Qcodes: docs/dataset/introduction.rst

- **Exporting datasets**: :doc:`Exporting data to other file formats <../examples/DataSet/Exporting-data-to-other-file-formats>`
- **Extracting runs between databases**: :doc:`Extracting runs from one DB file to another <../examples/DataSet/Extracting-runs-from-one-DB-file-to-another>` and :func:`qcodes.dataset.extract_runs_into_db`
- **Bulk export and metadata-only databases**: :func:`qcodes.dataset.export_datasets_and_create_metadata_db` for creating lightweight metadata-only databases while exporting all data to NetCDF files

.. _sec:intro_split_raw_data:

Split Raw Data Storage
======================

> By default, all measurement data (the results table rows) is stored in the same SQLite database alongside metadata such as experiments, runs, parameter layouts, and dependencies. Over time, the main database file can grow very large, which can slow down operations like browsing experiments and loading metadata.

Author: ```suggestion
By default, all measurement data (the results table rows) is stored in the same SQLite database alongside metadata such as experiments, runs, parameters and their dependencies. Over time, the main database file can grow very large, which can slow down operations like browsing experiments and make managing a large database file inconvenient.
```

    qc.config.dataset.raw_data_to_separate_db = True
    qc.config.dataset.raw_data_path = "/data/raw_measurements/"

If the per-dataset raw data files are moved to a different folder (e.g. during data migration or archival), the stored paths in the main database will become stale. Use the :func:`~qcodes.dataset.update_raw_data_paths` helper to update them::

    from qcodes.dataset import update_raw_data_paths

    update_raw_data_paths(
        db_path="/path/to/main_database.db",
        new_raw_data_folder="/new/location/of/raw_files/"
    )

This scans all datasets with a ``raw_data_db_path`` metadata entry, checks whether the corresponding ``.db`` file exists in the new folder, and updates the stored path accordingly.

Managing Datasets
-----------------

QCoDeS provides helpers for managing per-dataset raw data files over time:

- :func:`~qcodes.dataset.purge_orphaned_datasets` — remove dataset records whose raw data files no longer exist on disk.
- :func:`~qcodes.dataset.cleanup_datasets` — remove datasets (DB records and raw data files) by age, sample name, or file size.

For usage examples, see the :doc:`Database notebook <../examples/DataSet/Database>`.

> For more details on database management, see the :doc:`Database notebook <../examples/DataSet/Database>`.

Author: Addressed in ae3c879a6 - collapsed into a single sentence.

This scans all datasets with a ``raw_data_db_path`` metadata entry, checks whether the corresponding ``.db`` file exists in the new folder, and updates the stored path accordingly.

Purging Orphaned Datasets
-------------------------

After archiving or deleting selected raw data files, some dataset records in the main database may reference files that no longer exist on disk. Use :func:`~qcodes.dataset.purge_orphaned_datasets` to identify and remove those orphaned records::

    from qcodes.dataset import purge_orphaned_datasets

    # First, see what would be removed (dry run — the default)
    result = purge_orphaned_datasets("/path/to/main_database.db")
    print(f"{len(result.orphaned_datasets)} orphaned datasets found")

    # Then actually remove them
    result = purge_orphaned_datasets("/path/to/main_database.db", dry_run=False)

> Cleaning Up Datasets by Criteria

Author: please don't copy this information here, just make a reference to this same info in the database ipy notebook. as it's done above for exporting datasets and such.

Author: Addressed in c76baadf7.
